File: back-end/go/internal/mongodb/user_create.go
package mongodb

import (
	"context"
	"fmt"
	"github.com/icaroribeiro/hash-code-challenge/back-end/go/internal/models"
	"go.mongodb.org/mongo-driver/bson/primitive"
	"go.mongodb.org/mongo-driver/mongo"
)

func (d *Datastore) CreateUser(user models.User) (models.User, error) {
	var result *mongo.InsertOneResult
	var err error
	var id primitive.ObjectID
	var isOK bool

	result, err = d.DB.Collection("users").InsertOne(
		context.Background(),
		user,
	)

	if err != nil {
		return user, err
	}

	// Type assertion.
	id, isOK = result.InsertedID.(primitive.ObjectID)

	if isOK {
		user.ID = id
	} else {
		return user, fmt.Errorf("it wasn't possible to get the id of the generated document")
	}

	return user, nil
}

File: back-end/nodejs/internal/mongodb/user_create.js
var User = require('../models/user.js');

function CreateUser(user, callback) {
    User.create(user)
        .then((doc) => {
            if (doc) {
                callback(null, {
                        id: doc._id,
                        first_name: doc.first_name,
                        last_name: doc.last_name,
                        date_of_birth: doc.date_of_birth
                    }
                )
            } else {
                callback(null, null);
            }
        })
        .catch(err => {
            callback(err, null);
        });
}

exports.CreateUser = CreateUser;